### PR TITLE
Allow `None` when using a `Union` containing `Any` or `object`

### LINKED ISSUE #3444: "none is not an allowed value" when using Union with Any
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug
It seems that `None` is allowed when a field's type is `Any` (as is expected, stated in the docs), but `None` is not allowed when a field's type is `Union[SomeType, Any]`.

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.2
            pydantic compiled: True
                 install path: C:\Users\TobyHarradine\PycharmProjects\orchestration\.venv\Lib\site-packages\pydantic
               python version: 3.7.9 (tags/v3.7.9:13c94747c7, Aug 17 2020, 18:58:18) [MSC v.1900 64 bit (AMD64)]
                     platform: Windows-10-10.0.19041-SP0
     optional deps. installed: ['typing-extensions']
```

Code which reproduces the issue:
```pycon
>>> import pydantic
>>> class M(pydantic.BaseModel):
...     a: Any
...
>>> M(a=1)
M(a=1)
>>> M(a=None)
M(a=None)
>>> class M(pydantic.BaseModel):
...     a: Union[int, Any]
...
>>> M(a=1)
M(a=1)
>>> M(a="abcd")
M(a='abcd')
>>> M(a=None)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "pydantic\main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for M
a
  none is not an allowed value (type=type_error.none.not_allowed)

```

This may be somewhat related to #1624.

### PR DESCRIPTION
**Pull request by tharradine**
## Change Summary

This sets `allow_none` and `required` correctly when `Any` or `Object` are contained in a `Union` type.

## Related issue number

Resolves #3444.

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/3444-tharradine.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by tharradine** [comment]
Please review :)

File: pydantic/fields.py [hunk 1]
```diff
@@ -563,10 +563,11 @@ def _type_analysis(self) -> None:  # noqa: C901 (ignore complexity)
         if is_union_origin(origin):
             types_ = []
             for type_ in get_args(self.type_):
-                if type_ is NoneType:
+                if type_ is NoneType or type_ is Any or type_ is object:
```
Comment (PrettyWood) [inline pydantic/fields.py:566]:
```suggestion
                if is_none_type(type_) or type_ is Any or type_ is object:
```
same under

**Review by PrettyWood — COMMENTED** [review]
LGTM otherwise! Thanks

**Comment by PrettyWood** [comment]
please update

File: pydantic/fields.py (hunk 1 above)
Comment (Tobotimus) [reply pydantic/fields.py]: Fixed now, thanks!

**Comment by Tobotimus** [comment]
please review

File: pydantic/fields.py (hunk 1 above)
Comment (tharradine) [reply pydantic/fields.py]: Fixed now, thanks!

**Comment by tharradine** [comment]
please review

I accidentally commented in my personal account before, sorry for the spam!

**Review by PrettyWood — APPROVED** [review]

**Comment by samuelcolvin** [comment]
thanks so much.

**Pull request merged**